client/components/forms/form-text-input/index.jsx

```diff
@@ -20,11 +20,14 @@ export default class FormTextInput extends PureComponent {
 	constructor() {
 		super( ...arguments );
 
+		this.textField = React.createRef();
+		this.textFieldRef = null;
+
 		this.selectOnFocus = this.selectOnFocus.bind( this );
 	}
 
 	focus() {
-		this.refs.textField.focus();
+		this.textFieldRef.current.focus();
```

Comment:
What if we pass `inputRef` that's not an instance of `React.createRef`, but a simple callback or even a string? Then `this.textFieldRef` won't have a `current` field and calling `focus()` will crash.

Before this patch, the component had a similar bug where `focus()` could crash for another reason: if the component gets an `inputRef` prop, the `this.refs.textField` is never set and calling `focus()` on it fails. Maybe that's the reason why @ramonjd added a [very strange try/catch block](https://github.com/Automattic/wp-calypso/blob/fbd4229939d30569604676c34c3627d005379e78/client/components/domains/contact-details-form-fields/index.jsx#L248) around `focus` call in #19889?

One solution is to handle two refs: the internal and external one:
```js
class FormTextInput {
  textFieldRef = React.createRef();

  handleRef = node => {
    this.textFieldRef( node );
    this.props.inputRef && this.props.inputRef( node );
  }

  focus() {
    // textFieldRef is always set and always has the current field
    this.textFieldRef.current.focus();
  }

  render() {
    return <input ref={ this.handleRef } />;
  }
}
```

Then we can pass refs to `FormTextInput` in two ways:
```js
<FormTextInput ref={ textRef } />
```
Here the `textRef` callback will be passed an instance of the `FormTextInput` component. Because the component has a `focus` method, it can be called: `textRef.current.focus()`. But it doesn't have `blur`, so `textRef.current.blur()` is not possible.

```js
<FormTextInput inputRef={ textRef } />
```
Here the `textRef` callback gets a reference to the internal `<input>` DOM element. `textRef.current.focus()` will call directly the DOM method. `textRef.current.blur()` will work out of the box, too.

The `ref`s and `inputRef`s in our form elements and their usage in Domains would deserve a major cleanup that would take advantage of [React.forwardRef](https://reactjs.org/docs/forwarding-refs.html). If we used `React.forwardRef`, we could write:
```js
<FormTextInput ref={ textRef } />
```
and the `textRef` would point to the internal `<input>` element -- something that we probably always want anyway. No need to discriminate between `ref` and `inputRef` anywhere and no need for [code like this](https://github.com/Automattic/wp-calypso/blob/fbd4229939d30569604676c34c3627d005379e78/client/components/domains/contact-details-form-fields/index.jsx#L320-L322).

Maybe a good todo item for the Muriel Components project? (cc @blowery @griffbrad)

Reply:
@jsnajdr 
Thanks for such a clear, actionable and detailed feedback. 
I will make the changes following your suggested solution.

It seems that a major cleanup with `React.forwardRef` would require converting the `FormTextInput` component to a function or a higher-order-component, since regular class components don't receive `ref` as arguments and `ref` is not available in `props`. Is that the case?

Reply:
@jsnajdr 
For the use case `<FormTextInput ref={ textRef } />`,  should I add a `blur` method to the `FormTextInput` component to cater for `textRef.current.blur()`?

Reply:
> Maybe that's the reason why @ramonjd added a very strange try/catch block around focus call in #19889?

👍 You got me @jsnajdr :) That's exactly why. It was pre `16.3` and we were mud wrestling with `form-state`. Dreaming about building state-driven forms using Muriel Components... 🐑